Add `AGENTS.md` and AI contribution guidance (#2193)
WordPress/performance: AGENTS.md

# Performance Lab

This is a monorepo for the WordPress Performance Team, containing a collection of standalone performance feature plugins. Refer to the [Performance Lab handbook](https://make.wordpress.org/performance/handbook/performance-lab/) for more details. 

## Project Overview

> * **Purpose:** To develop and maintain a suite of plugins that improve the performance of WordPress sites. All should be considered 

Reviewer: > All should be considered 

Was there supposed to be something else here?

Author: Oops.

```suggestion
* **Purpose:** To develop and maintain a suite of plugins that improve the performance of WordPress sites. All should be considered potential for future candidates for merging into WordPress core.
```

* **Technologies:** PHP, JavaScript, CSS, a variety of testing and linting tools.

> * **Architecture:** This is a monorepo containing multiple WordPress plugins. The main plugin is `performance-lab`, which acts as a hub for the other feature plugins, most of which are standalone plugins located in the `plugins` directory (although some come from other repos). Each plugin has its own tests located in its respective `tests` subdirectory.

Reviewer 2: Collection of standalone performance features instead of plugin?

Author: Like this?

```suggestion
* **Architecture:** This is a monorepo containing multiple WordPress plugins. The main plugin is `performance-lab`, which acts as a hub for performance features, most of which are standalone plugins located in the `plugins` directory (although some come from other repos). Each plugin has its own tests located in its respective `tests` subdirectory.
```

## Building and Running

### Prerequisites

* [Node.js and npm](https://nodejs.org/en/)
* [Docker](https://www.docker.com/)
* [Git](https://git-scm.com/)
* [Composer](https://getcomposer.org/)

### Installation

1. Clone the repository, e.g. `git clone https://github.com/WordPress/performance.git`.
2. Run `npm install` to install the Node.js dependencies.

> 3. Run `composer install` to install the PHP dependencies.

Reviewer 3: How is this relevant for the LLM / agent? It's going to work within the repo, so I'm not sure about the value about adding this information here.

Author: Gemini CLI included it when doing `/init` and Gutenberg's `AGENTS.md` also includes it: https://github.com/WordPress/gutenberg/pull/71796

Reviewer 3: Where exactly? I don't see it including how to clone the repository, which seems pointless to me.

Author: I didn't know you were commenting specifically about that line. But Gemini CLI added that as the initial step for the initial `GEMINI.md`. I'm happy to remove.

Reviewer 3: I wasn't, my overall feedback is I would phrase it more closely to how Gutenberg does it: It's not that the LLM has to perform these steps to install. It's probably more of a "good to know" for how the project is installed, but in practice the project will probably already be installed in 99% of LLM use-cases.

Reviewer 3: Personally I'm not sure of the value of having this at all, but that's only a hunch that I can't back up with evidence. But so is likely almost anything we add to these `AGENTS.md` files, unless we run some actual evals on different approaches.

Author: See 6dd5c7f5890c48733cc296ef4eddf0643a1809a7

### Building

> * To build the JavaScript and CSS assets: `npm run build``

Reviewer 3: Typo / formatting error:
```suggestion
* To build the JavaScript and CSS assets: `npm run build`
```

* To build all plugins and place into the `build` directory: `npm run build-plugins`
* To build a specific plugin: `npm run build:plugin:<plugin-slug>` (e.g., `npm run build:plugin:performance-lab`)
* To build ZIP files for distribution: `npm run build-plugins:zip`

### Running a Local Environment

This project uses `@wordpress/env` to create a local development environment.

* Check if the environment is already running: `npm run wp-env status`
* Start the environment: `npm run wp-env start`
* Stop the environment: `npm run wp-env stop`

The environment will by default be located at `http://localhost:8888` but this can be overridden by `.wp-env.override.json`.

> ## Code Style

Reviewer 3: Just an idea: Consider adding a PHP code snippet and a JavaScript code snippet with some very simple demo function each to showcase the correct code style and documentation style. Those examples are often a better way for LLMs to know what to follow rather than detailed prose explanations.

It also allows bringing it all together, since you can cover both code style and documentation formatting (3P-verb descriptions, `@param` and `@return` in one group, `@since n.e.x.t` requirement, etc.).

Author: I added a couple code snippets: 385a6be3b9d612e2b228086a42c29b5885d56005

Reviewer 3: Looks great!

For the HTML Coding Standards, disregard the guidance that void/empty tags should be self-closing, such as `IMG`, `BR`, `LINK`, or `META`. This is only relevant for XML (XHTML), not HTML. So instead of `<br />` this should only use `<br>`, for example.

Additionally, the [inline documentation standards for WordPress](https://developer.wordpress.org/coding-standards/inline-documentation-standards/) should be followed:

* [PHP Documentation Standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/)
* [JavaScript Documentation Standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/)

Note that `lint-staged` will be used to automatically run code quality checks with the tooling based on the staged files.

### Indentation

In general, indentation should use tabs. Refer to `.editorconfig` in the project root for specifics.

### Inline Documentation

> It is expected for code introduced in pull requests to have `@since` tags with the `n.e.x.t` placeholder version. It will get replaced with the actual version at the time of release. Do not add any code review comments such code.

Reviewer: ```suggestion
It is expected for code introduced in pull requests to have `@since` tags with the `n.e.x.t` placeholder version. It will get replaced with the actual version at the time of release. Do not add any code review comments to such code.
```

> It is expected for code introduced in pull requests to have `@since` tags with the `n.e.x.t` placeholder version. It will get replaced with the actual version at the time of release. Do not add any code review comments such code.

Reviewer 3: Doesn't really matter whether it's introduced in pull requests or not. That's the norm, but it might also confuse the LLM ("am I currently working on a pull request or just a local branch?").

```suggestion
It is expected for new code introduced to have `@since` tags with the `n.e.x.t` placeholder version. It will get replaced with the actual version at the time of release. Do not add any code review comments such code.
```

# Performance Lab

This is a monorepo for the WordPress Performance Team, containing a collection of standalone performance feature plugins. Refer to the [Performance Lab handbook](https://make.wordpress.org/performance/handbook/performance-lab/) for more details. 

## Project Overview

* **Purpose:** To develop and maintain a suite of plugins that improve the performance of WordPress sites. All should be considered potential for future candidates for merging into WordPress core.
* **Technologies:** PHP, JavaScript, CSS, a variety of testing and linting tools.

> * **Architecture:** This is a monorepo containing multiple WordPress plugins. The main plugin is `performance-lab`, which acts as a hub for performance features, most of which are standalone plugins located in the `plugins` directory (although some come from other repos). Each plugin has its own tests located in its respective `tests` subdirectory.

Reviewer 3: Consider replacing this with a more structured list of key folders/files, e.g. in a `### Project Structure` sub section.

Something like:
```suggestion

### Project Structure

* `/bin`: Custom CLI commands and scripts for certain development workflows.
* `/plugins`: The actual WordPress plugins that are developed in this monorepo.
* `/plugins/*`: An individual WordPress plugin folder.
* `/plugins/*/tests`: PHPUnit tests for the specific WordPress plugin.
* `/tools`: Setup and configuration files for various tools, such as linting and testing.
```

* To build a specific plugin: `npm run build:plugin:<plugin-slug>` (e.g., `npm run build:plugin:performance-lab`)
* To build ZIP files for distribution: `npm run build-plugins:zip`

### Running a Local Environment

This project uses `@wordpress/env` to create a local development environment.

* Check if the environment is already running: `npm run wp-env status`
* Start the environment: `npm run wp-env start`
* Stop the environment: `npm run wp-env stop`

The environment will by default be located at `http://localhost:8888` but this can be overridden by `.wp-env.override.json`.

## Code Style

In general, the [coding standards for WordPress](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/) should be followed:

* [CSS Coding Standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/css/)
* [HTML Coding Standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/html/)
* [JavaScript Coding Standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/)
* [PHP Coding Standards](https://developer.wordpress.org/coding-standards/wordpress-coding-standards/php/)

> Note that for the JavaScript Coding Standards, the code should also be formatted using Prettier, specifically the [wp-prettier](https://www.npmjs.com/package/wp-prettier) fork with the `--paren-spacing` option which inserts many extra spaces inside parentheses.

Reviewer 4: ```suggestion
Note that for the JavaScript Coding Standards, the code should also be formatted using Prettier, specifically the [wp-prettier](https://www.npmjs.com/package/wp-prettier) fork with the `--paren-spacing` option which inserts extra spaces inside parentheses.
```

For the HTML Coding Standards, disregard the guidance that void/empty tags should be self-closing, such as `IMG`, `BR`, `LINK`, or `META`. This is only relevant for XML (XHTML), not HTML. So instead of `<br />` this should only use `<br>`, for example.

Additionally, the [inline documentation standards for WordPress](https://developer.wordpress.org/coding-standards/inline-documentation-standards/) should be followed:

* [PHP Documentation Standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/)
* [JavaScript Documentation Standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/)

Note that `lint-staged` will be used to automatically run code quality checks with the tooling based on the staged files.

### Indentation

In general, indentation should use tabs. Refer to `.editorconfig` in the project root for specifics.

### Inline Documentation

> It is expected for new code introduced to have `@since` tags with the `n.e.x.t` placeholder version. It will get replaced with the actual version at the time of release. Do not add any code review comments such to code.

Reviewer 3: Typo I believe?
```suggestion
It is expected for new code introduced to have `@since` tags with the `n.e.x.t` placeholder version. It will get replaced with the actual version at the time of release. Do not add any code review comments to such code.
```

Every file, function, class, method constant, and global variable must have an associated docblock with a `@since` tag.

### PHP

Follow coding conventions in WordPress core. Namespaces are generally not used, as they are not normally used in WordPress core code. Procedural programming patterns are favored where classes play a supporting role, rather than everything being written in OOP.

> Whenever possible, the most specific PHP type hints should be used, when backward compatible with PHP 7.2, the minimum version of PHP supported by WordPress and this repository. When native PHP type cannot be used, PHPStan's [PHPDoc Types](https://phpstan.org/writing-php-code/phpdoc-types) should be used, including not only the basic types but also subtypes like `non-empty-string`, [integer ranges](https://phpstan.org/writing-php-code/phpdoc-types#integer-ranges), [general arrays](https://phpstan.org/writing-php-code/phpdoc-types#general-arrays), and especially [array shapes](https://phpstan.org/writing-php-code/phpdoc-types#array-shapes). The types should comply with PHPStan's level 10. The one exception for using PHP types is whenever a function is used as a filter. Since plugins can supply any value at all when filtering, use the expected type with a union to `mixed`. The first statement in the function in this case must always check the type, and if it is not the expected type, override it to be so.

Reviewer 4: note this might soon become 7.4, see https://core.trac.wordpress.org/ticket/62622#comment:37

maybe a link to the place where we list the minimum version required for wordpress and the plugin?

Reviewer 3: See https://github.com/WordPress/performance/pull/2193#discussion_r2392087473, while the duplication aspect is annoying, I think inlining is more effective with LLMs than pointing them to read another file.

Reviewer 4: Sounds good, just a mental note then to update this!

> Whenever possible, the most specific PHP type hints should be used, when compatible with the minimum version of PHP supported by WordPress, unless the `testVersion` config in `phpcs.xml.dist` is higher. When native PHP type cannot be used, PHPStan's [PHPDoc Types](https://phpstan.org/writing-php-code/phpdoc-types) should be used, including not only the basic types but also subtypes like `non-empty-string`, [integer ranges](https://phpstan.org/writing-php-code/phpdoc-types#integer-ranges), [general arrays](https://phpstan.org/writing-php-code/phpdoc-types#general-arrays), and especially [array shapes](https://phpstan.org/writing-php-code/phpdoc-types#array-shapes). The types should comply with PHPStan's level 10. The one exception for using PHP types is whenever a function is used as a filter. Since plugins can supply any value at all when filtering, use the expected type with a union to `mixed`. The first statement in the function in this case must always check the type, and if it is not the expected type, override it to be so.

Reviewer 3: I assume you wanted to avoid duplication, but it's usually better to include information inline instead of referring to different files. So I think it's better to simply put `7.2` in here. It's one more place to update, but at the same time it's also very easy to do a search for `7.2` throughout the codebase whenever we bump the minimum requirement, which FWIW also happens so rarely this little overhead shouldn't matter.
```suggestion
Whenever possible, the most specific PHP type hints should be used, when backward compatible with PHP 7.2, the minimum version of PHP supported by WordPress and this repository. When native PHP type cannot be used, PHPStan's [PHPDoc Types](https://phpstan.org/writing-php-code/phpdoc-types) should be used, including not only the basic types but also subtypes like `non-empty-string`, [integer ranges](https://phpstan.org/writing-php-code/phpdoc-types#integer-ranges), [general arrays](https://phpstan.org/writing-php-code/phpdoc-types#general-arrays), and especially [array shapes](https://phpstan.org/writing-php-code/phpdoc-types#array-shapes). The types should comply with PHPStan's level 10. The one exception for using PHP types is whenever a function is used as a filter. Since plugins can supply any value at all when filtering, use the expected type with a union to `mixed`. The first statement in the function in this case must always check the type, and if it is not the expected type, override it to be so.
```
